BPA/Ch.09/ladyslipper.stan

```diff
@@ -0,0 +1,126 @@
+// -------------------------------------------------
+// States (S):
+// 1 vegetative
+// 2 flowering
+// 3 dormant
+// 4 dead
+// Observations (O):  
+// 1 seen vegetative 
+// 2 seen flowering
+// 3 not seen
+// -------------------------------------------------
+
+functions {
+  int first_capture(int[] y_i) {
+    for (k in 1:size(y_i))
+      if (y_i[k] != 3)
+        return k;
+    return 0;
+  }
+}
+
+data {
+  int<lower=0> nind;
+  int<lower=0> n_occasions;
+  int<lower=1,upper=3> y[nind, n_occasions];
+}
+
+transformed data {
+  int<lower=0,upper=n_occasions> first[nind];
+
+  for (i in 1:nind)
+    first[i] <- first_capture(y[i]);
+}
+
+parameters {
+  vector<lower=0,upper=1>[n_occasions-1] s; // Survival probability
+  vector<lower=0>[3] a;
+  vector<lower=0>[3] b;
+  vector<lower=0>[3] c;
+}
+
+transformed parameters {
+  simplex[3] psiD; // Transitions from dormant
+  simplex[3] psiV; // Transitions from vegetative
+  simplex[3] psiF; // Transitions from flowering
+  simplex[4] ps[4, nind, n_occasions-1];
+  simplex[3] po[4, nind, n_occasions-1];
+
+  // Constraints
+  for (i in 1:3){
+    psiD[i] <- a[i] / sum(a);
+    psiV[i] <- b[i] / sum(b);
+    psiF[i] <- c[i] / sum(c);
+  }
+
+  // Define state-transition and observation matrices 	
+  for (i in 1:nind) {
+    // Define probabilities of state S(t+1) given S(t)
+    for (t in 1:(n_occasions - 1)) {
+      ps[1, i, t, 1] <- s[t] * psiV[1];
+      ps[1, i, t, 2] <- s[t] * psiV[2];
+      ps[1, i, t, 3] <- s[t] * psiV[3];
+      ps[1, i, t, 4] <- 1.0 - s[t];
+      ps[2, i, t, 1] <- s[t] * psiF[1];
+      ps[2, i, t, 2] <- s[t] * psiF[2];
+      ps[2, i, t, 3] <- s[t] * psiF[3];
+      ps[2, i, t, 4] <- 1.0 - s[t];
+      ps[3, i, t, 1] <- s[t] * psiD[1];
+      ps[3, i, t, 2] <- s[t] * psiD[2];
+      ps[3, i, t, 3] <- s[t] * psiD[3];
+      ps[3, i, t, 4] <- 1.0 - s[t];
+      ps[4, i, t, 1] <- 0.0;
+      ps[4, i, t, 2] <- 0.0;
+      ps[4, i, t, 3] <- 0.0;
+      ps[4, i, t, 4] <- 1.0;
+
+      // Define probabilities of O(t) given S(t)
+      po[1, i, t, 1] <- 1.0;
```

Comment:
po should be defined as transformed data.  And if you could manage it somehow, so should ps[4].  

And you can use the new syntax to do something like  

```
po[,i,t,] <- diag_matrix(rep_vector(1, 3));
```

but it'd be even better to just define a single unit matrix and reuse it (that is, drop the middle indices, which don't do anything).